teuthology/provision/cloud/openstack.py

```diff
@@ -0,0 +1,331 @@
+import logging
+import requests
+import socket
+import time
+import urllib
+import yaml
+
+from copy import deepcopy
+
+from paramiko import AuthenticationException
+from paramiko.ssh_exception import NoValidConnectionsError
+
+from teuthology.config import config
+from teuthology.contextutil import safe_while
+
+import base
+import util
+from teuthology.provision.cloud.base import Provider
+
+
+log = logging.getLogger(__name__)
+
+
+class OpenStackProvider(Provider):
+    _driver_posargs = ['username', 'password']
+
+    @property
+    def images(self):
+        if not hasattr(self, '_images'):
+            self._images = self.driver.list_images()
+        return self._images
+
+    @property
+    def sizes(self):
+        if not hasattr(self, '_sizes'):
+            self._sizes = self.driver.list_sizes()
+        return self._sizes
+
+    @property
+    def networks(self):
+        if not hasattr(self, '_networks'):
+            try:
+                self._networks = self.driver.ex_list_networks()
+            except AttributeError:
+                log.warn("Unable to list networks for %s", self.driver)
+                self._networks = list()
+        return self._networks
+
+    @property
+    def security_groups(self):
+        if not hasattr(self, '_security_groups'):
+            try:
+                self._security_groups = self.driver.ex_list_security_groups()
+            except AttributeError:
+                log.warn("Unable to list security groups for %s", self.driver)
+                self._security_groups = list()
+        return self._security_groups
+
+
+class OpenStackProvisioner(base.Provisioner):
+    _sentinel_path = '/.teuth_provisioned'
+
+    defaults = dict(
+        openstack=dict(
+            machine=dict(
+                disk=20,
+                ram=8000,
+                cpus=1,
+            ),
+            volumes=dict(
+                count=0,
+                size=0,
+            ),
+        )
+    )
+
+    def __init__(
+        self,
+        provider, name, os_type=None, os_version=None,
+        conf=None,
+        user='ubuntu',
+    ):
+        super(OpenStackProvisioner, self).__init__(
+            provider, name, os_type, os_version, conf=conf, user=user,
+        )
+        self._read_conf(conf)
+
+    def _read_conf(self, conf=None):
+        """
+        Looks through the following in order:
+
+            the 'conf' arg
+            conf[DRIVER_NAME]
+            teuthology.config.config.DRIVER_NAME
+            self.defaults[DRIVER_NAME]
+
+        For each, looks for sections named 'machine' or 'volume. The first one
+        found of each type is used to configure nodes with disk, RAM, cpu,
+        volume size and count
```

Comment:
it looks like combine_dicts with that lambda will actually use the max of all sources?...